### DISCUSSION TITLE
save view

### DISCUSSION THREAD

**Discussion by paprophet** [iTwin/itwinjs-core#286] [Q&A] [answered]
i use web page from imodeljs sample (APPUI Viewport Frontstage)  by imodeljs ，and i see the view list (IModelConnectedViewSelector);
how i can  create  new view into viewlist by imodeljs

**Comment by shenshengyi** [comment, marked as answer]
ViewCreator2d,ViewCreator3d.
Using these two interfaces, you can create a new 2D or 3D view in the front end,

**Reply by paprophet** [reply]
> ViewCreator2d,ViewCreator3d.
> Using these two interfaces, you can create a new 2D or 3D view in the front end,
for example i  rotate the model some degrees and clip the model in the web ，how i save the current some parameter to save new view,and next i open the web,i can load the save view to see (have i save the  degrees value and clip)
how i can do to implementation requirements by  ViewCreator3d or other method?

**Reply by shenshengyi** [reply]
> > ViewCreator2d,ViewCreator3d.
> > Using these two interfaces, you can create a new 2D or 3D view in the front end,
> > for example i  rotate the model some degrees and clip the model in the web ，how i save the current some parameter to save new view,and next i open the web,i can load the save view to see (have i save the  degrees value and clip)
> > how i can do to implementation requirements by  ViewCreator3d or other method?

Hi, you can refer to the following code to serialize the current view property into a string data, and then save it to the external by implementing RPC.

Or use IModelApp.settings The interface saves the data to the server, and the next time it opens, it gets the data and deserializes it to viewstate, and then sets it to the current view.

I use it in this way at present, and it works very well.

export async function TestSerializationView() {
  const vp = IModelApp.viewManager.selectedView!.view;
  const viewProp = SavedView.viewStateToProps(vp);
  const strViewProp = JSON.stringify(viewProp);
  const savedViewFilePath = Config.App.get("imjs_savedview_file");
  await SVTRpcInterface.getClient().writeExternalSavedViews(
    savedViewFilePath,
    strViewProp
  );
}
export async function TestDeSerializationView() {
  const savedViewFilePath = Config.App.get("imjs_savedview_file");
  const strViewProp = await SVTRpcInterface.getClient().readExternalSavedViews(
    savedViewFilePath
  );
  const vp = IModelApp.viewManager.selectedView!;
  const viewProp: SavedViewProps = JSON.parse(strViewProp);
  const imodel = UiFramework.getIModelConnection()!;
  const viewState = await SavedView.viewStateFromProps(imodel, viewProp);
  if (viewState) {
    vp.changeView(viewState);
  }
}

**Reply by paprophet** [reply]
> > > ViewCreator2d,ViewCreator3d.
> > > Using these two interfaces, you can create a new 2D or 3D view in the front end,
> > > for example i  rotate the model some degrees and clip the model in the web ，how i save the current some parameter to save new view,and next i open the web,i can load the save view to see (have i save the  degrees value and clip)
> > > how i can do to implementation requirements by  ViewCreator3d or other method?
> 
> Hi, you can refer to the following code to serialize the current view property into a string data, and then save it to the external by implementing RPC.
> 
> Or use IModelApp.settings The interface saves the data to the server, and the next time it opens, it gets the data and deserializes it to viewstate, and then sets it to the current view.
> 
> I use it in this way at present, and it works very well.
> 
> export async function TestSerializationView() {
> const vp = IModelApp.viewManager.selectedView!.view;
> const viewProp = SavedView.viewStateToProps(vp);
> const strViewProp = JSON.stringify(viewProp);
> const savedViewFilePath = Config.App.get("imjs_savedview_file");
> await SVTRpcInterface.getClient().writeExternalSavedViews(
> savedViewFilePath,
> strViewProp
> );
> }
> export async function TestDeSerializationView() {
> const savedViewFilePath = Config.App.get("imjs_savedview_file");
> const strViewProp = await SVTRpcInterface.getClient().readExternalSavedViews(
> savedViewFilePath
> );
> const vp = IModelApp.viewManager.selectedView!;
> const viewProp: SavedViewProps = JSON.parse(strViewProp);
> const imodel = UiFramework.getIModelConnection()!;
> const viewState = await SavedView.viewStateFromProps(imodel, viewProp);
> if (viewState) {
> vp.changeView(viewState);
> }
> }

thank you!i use "SavedView.viewStateToProps" and  solved this problem,thank you!

**Comment by shenshengyi** [comment]
In fact, it mainly uses the interface of savedview.
